### DISCUSSION TITLE
How can I set custom HTTP Headers

### DISCUSSION THREAD

**Discussion by apotek** [maennchen/ZipStream-PHP#305] [Q&A] [answered]
Hi, for a variety of reasons, I am still (stuck) on version 2.4.  I wanted to set a custom Cache-Control header, since there is no method for setting such a header. I looked at the Archive class (through which I can set options) and there is a method  on it called `setHttpHeaderCallback()`, which seems to accept a callable.  When I look at the class docs for the 2.4 version of the Archive class, I see the method documented, but there is no documentation of what should be returned by my callable. 

I tried this:

```php
$zip_options = new ArchiveOptions(); // Imported as ArchiveOptions for legibility
...
...
...
$zip_options->setHttpHeaderCallback([$this, 'zipHeaders']);
```

And my `zipHeaders()` method returned an array like this:

```php
public function zipHeaders() {
  return [
    'header-one: header one value',
    'another-header: value 2',
  ];
}
```
I  also implemented this function returning not an array but a `\n` delimited string. 

In any event, neither of my attempts worked and none of my test headers were returned. I did trace it with x-debug and the callback was called (five times, which was interesting), but the headers were missing from the response.

What kind of return value should be returned from the callable?

Thank you.

ps., If I am completely missing the boat and there is another way to set a  `Cache-Control:...` header, I am all ears

**Comment by maennchen** [comment, marked as answer]
@apotek The header callback is used to actually set the header. Its arguments are the header name and the value. No return is expected.

Therefore this is not the functionality you need.

Instead, just disable the headers in ZipStream and set the headers yourself. See:
* https://github.com/maennchen/ZipStream-PHP/blob/620c2aab9d4f109aef4949049cc909c24479f8e0/src/ZipStream.php#L518-L544
* https://github.com/maennchen/ZipStream-PHP/blob/620c2aab9d4f109aef4949049cc909c24479f8e0/src/Option/Archive.php#L42-L48

**Comment by apotek** [comment]
Thank you for your reply. I am probably misunderstanding, but it seems that the instructions are then to:

```php
foreach ($headers as $header => $value) {
  my_preferred_header_function_for_my_framework($header . ': ' . $value);
}

$zip->finish();
```

My experience with trying to send my own headers seemed to cause the zip to not download as a file (despite setting content disposition and filename and mimetype etc). I'll look at that again. 

So the instructions for 2.4 are:
1. disable sending http headers (this is default so just make sure you aren't enabling it).
2. Send your own headers however you normally would do in your framework (or via raw php function `header()`)
3. Finish the zip stream.

Is that correct? Thanks again.

**Reply by maennchen** [reply]
I would send the headers before you add the first file. Otherwise you’ll already be streaming the content and can no longer set headers.

The rest makes sense 👍
